webapp/channels/src/packages/mattermost-redux/src/selectors/entities/channel_categories.ts

```diff
@@ -447,13 +447,18 @@ export function makeGetChannelsByCategory() {
 
         const channelsByCategory: RelationOneToOne<ChannelCategory, Channel[]> = {};
 
+        // TODO: This avoids some rendering, but there is a bigger issue underneath
+        // Every time myPreferences or myChannels change (which can happen for many
+        // unrelated reasons) the whole list of channels gets reordered and re-filtered.
+        let allEquals = categoryIds === lastCategoryIds;
         for (const category of categories) {
             const channels = getChannels[category.id](state, category.channel_ids);
             channelsByCategory[category.id] = filterAndSortChannels[category.id](state, channels, category);
+            allEquals = allEquals && shallowEquals(channelsByCategory[category.id], lastChannelsByCategory[category.id]);
         }
 
         // Do a shallow equality check of channelsByCategory to avoid returning a new object containing the same data
-        if (shallowEquals(channelsByCategory, lastChannelsByCategory)) {
+        if (allEquals) {
```

Comment:
I'm not sure I understand the improvement here. It seems like we're just manually doing the same thing that calling `shallowEquals` would do, so I'm not sure how that helps reduce rendering

Reply:
The problem is in `filterAndSortChannels`. This is returning a different array of channels more often than expected. Therefore, if we do the shallow equals... even if the channels inside the category are the same, the arrays themselves are different, triggering a re-render.

This check each category one by one to verify they have the same channels, instead of verifying that the arrays are the same.